### ISSUE TITLE
bind(), call(), and apply() are untyped

### ISSUE THREAD

**Issue by jameslong** [Microsoft/TypeScript#212]
bind() returns type 'any' and does not type check it's arguments. e.g.

    var add = function (a: number, b: number) { return a + b; };
    var addString = add.bind(null, 'hello');

compiles without error and 'addString' has type any. 

bind is a default way to achieve partial application so it would be great to have it typechecked properly.

Similarly for apply, the code

    var listA = [1, 2, 3];
    var listB = ['a', 'b', 'c'];
    listA.push.apply(listA, listB);

compiles without error and listA is still incorrectly marked as type 'number[]'.

call() example:

    var add5 = function (a: number) { return 5 + a; };
    add5.call(null, 'a string');

Again without a compiler error.

**Comment by impinball** [comment]
`apply` and `call` could be easily fixed by the many variadic argument proposals (syntax only for demonstration, requires `this` typing).

```ts
interface Function {
  apply<T, R, ...X>(
    this: (this: T, ...args: X) => R,
    thisArg: T,
    args: X
  ): R;

  call<T, R, ...X>(
    this: (this: T, ...args: X) => R,
    thisArg: T,
    ...args: X
  ): R;
}
```

`bind` would be a beast to properly type, though, as a compiler would have to properly infer this kind of thing (syntax only for demonstration, requires `this` typing for the function version):

```ts
interface Function {
  // bound functions
  bind<T, R, ...X, ...Y>(
    this: (this: T, ...initial: X, ...rest: Y) => R,
    thisArg: T,
    ...initial: X
  ): (...rest: Y) => R;

  // bound constructors
  bind<C, ...X, ...Y>(
    this: new (...initial: X, ...rest: Y) => C,
    thisArg: any,
    ...initial: X
  ): new (...rest: Y) => C;
}
```

Let's just say it requires lazy inference of the template each use, probably using a simplified structure along the same lines as [Conway's Game of Life](https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life) or [Rule 110](https://en.wikipedia.org/wiki/Rule_110), something not currently done, and something not easy to program.

```ts
let bar: T;
let baz: A;
let quux: B;
let spam: C;
let eggs: D;
declare function foo(this: T, a: A, b: B, c: C, d: D): E;

let res = foo.bind(bar, baz, quux);
res(spam, eggs)
```

*Expansion/substitution of last statement:*

```ts
// Initial
bind<T, R, ...X, ...Y>(this: (this: T, ...initial: X, ...rest: Y) => R, thisArg: T, ...initial: X): (this: any, ...rest: Y) => R;

// Steps to reduce
bind<T, R,       ...X,       ...Y>(this: (this: T,             ...initial: X,             ...rest: Y) => R, thisArg: T,             ...initial: X): (            ...rest: Y) => R;
bind<T, E,       ...X,       ...Y>(this: (this: T,             ...initial: X,             ...rest: Y) => E, thisArg: T,             ...initial: X): (            ...rest: Y) => E;
bind<T, E, A,    ...X,       ...Y>(this: (this: T, a: A,       ...initial: X,             ...rest: Y) => E, thisArg: T, a: A,       ...initial: X): (            ...rest: Y) => E;
bind<T, E, A, B, ...X,       ...Y>(this: (this: T, a: A, b: B, ...initial: X,             ...rest: Y) => E, thisArg: T, a: A, b: B, ...initial: X): (            ...rest: Y) => E;
bind<T, E, A, B,             ...Y>(this: (this: T, a: A, b: B,                            ...rest: Y) => E, thisArg: T, a: A, b: B               ): (            ...rest: Y) => E;
bind<T, E, A, B,       C,    ...Y>(this: (this: T, a: A, b: B,                c: C,       ...rest: Y) => E, thisArg: T, a: A, b: B               ): (c: C,       ...rest: Y) => E;
bind<T, E, A, B,       C, D, ...Y>(this: (this: T, a: A, b: B,                c: C, d: D, ...rest: Y) => E, thisArg: T, a: A, b: B               ): (c: C, d: D, ...rest: Y) => E;
bind<T, E, A, B,       C, D      >(this: (this: T, a: A, b: B,                c: C, d: D            ) => E, thisArg: T, a: A, b: B               ): (c: C, d: D            ) => E;

// Signature of what's finally type checked
bind<T, E, A, B, C, D>(this: (this: T, A, B, C, D) => E, T, A, B): (C, D) => E;
```

Not simple. (Not even C++ can do this without major template hacks)

---

*Or, in highly technical CS jargon, checking the above type correctly may mean controlling the variadic expansion in each occurence via a simple [Class 1](https://en.wikipedia.org/wiki/Cellular_automaton#Classification) cellular automaton, reducing it until the variadic types have been eliminated. There is the possibility that a template expands infinitely, especially if the function is recursive (singly or cooperatively), but a hard limit could prevent a lock-up for the compiler.*

**Comment by impinball** [comment]
And as for this specific issue, it's nearly completely covered by the variadic generic proposal in #1773 and the `this` typing proposal in #229 and the large number of other dupes.

**Comment by mhegazy** [comment]
related to #3694

**Comment by wernerdegroot** [comment]
For those of you who care, I've create the following little TypeScript-snippet that outputs an overloaded function declaration of 'bind':

Output (you can scale this up if you want):
```
function bind<A, B, Z>(f: (_0: A, _1: B) => Z, _0: A): (_0: B) => Z;
function bind<A, B, C, Z>(f: (_0: A, _1: B, _2: C) => Z, _0: A): (_0: B, _1: C) => Z;
function bind<A, B, C, Z>(f: (_0: A, _1: B, _2: C) => Z, _0: A, _1: B): (_0: C) => Z;
function bind<A, B, C, D, Z>(f: (_0: A, _1: B, _2: C, _3: D) => Z, _0: A): (_0: B, _1: C, _2: D) => Z;
function bind<A, B, C, D, Z>(f: (_0: A, _1: B, _2: C, _3: D) => Z, _0: A, _1: B): (_0: C, _1: D) => Z;
function bind<A, B, C, D, Z>(f: (_0: A, _1: B, _2: C, _3: D) => Z, _0: A, _1: B, _2: C): (_0: D) => Z;
function bind<A, B, C, D, E, Z>(f: (_0: A, _1: B, _2: C, _3: D, _4: E) => Z, _0: A): (_0: B, _1: C, _2: D, _3: E) => Z;
function bind<A, B, C, D, E, Z>(f: (_0: A, _1: B, _2: C, _3: D, _4: E) => Z, _0: A, _1: B): (_0: C, _1: D, _2: E) => Z;
function bind<A, B, C, D, E, Z>(f: (_0: A, _1: B, _2: C, _3: D, _4: E) => Z, _0: A, _1: B, _2: C): (_0: D, _1: E) => Z;
function bind<A, B, C, D, E, Z>(f: (_0: A, _1: B, _2: C, _3: D, _4: E) => Z, _0: A, _1: B, _2: C, _3: D): (_0: E) => Z;
function bind(f, ...toBind: any[]) {
    return f.bind.apply(f, [null].concat(toBind));
}
```

Snippet:

```
function range(start: string,stop: string): string[] {
	var result: string[] = [];
	
	var index: number;
	for (var index = start.charCodeAt(0), end = stop.charCodeAt(0); index <= end; ++index){
		result.push(String.fromCharCode(index));
	}
	return result;
};

function untilIndex<T>(index: number, ts: T[]): T[] {
	return ts.slice(0, index);
}

function fromIndex<T>(index: number, ts: T[]): T[] {
	return ts.slice(index);
}

function getParameters(genericTypeNames: string[]): string[] {
	return genericTypeNames.map((genericTypeName, index) => {
		return '_' + index + ': ' + genericTypeName;
	});
}

function getParametersAsString(genericTypeNames: string[]): string {
	return getParameters(genericTypeNames).join(', ');
}

function getFunctionSignature(genericTypeNames: string[], returnTypeName: string): string {
	return '(' + getParametersAsString(genericTypeNames) + ') => ' + returnTypeName;
}

function getBindOverloads(genericTypeNames: string[]): string {
	
	var returnTypeName = 'Z';
	
	var functionToBindSignature: string = getFunctionSignature(genericTypeNames, returnTypeName);
	
	var overloads: string = '';
	
	var index: number;
	for (index = 1; index < genericTypeNames.length; ++index) {
		var before: string[] = untilIndex(index, genericTypeNames);
		var after: string[] = fromIndex(index, genericTypeNames);
		
		var hasBoundParameters: boolean = before.length > 0;
		var boundParametersAsString: string = getParametersAsString(before);
		var boundFunctionTypeAsString: string = getFunctionSignature(after, returnTypeName);
		
		overloads += 'function bind&lt;' + genericTypeNames.concat([returnTypeName]).join(', ') + '&gt;(' + 
			'f: ' + functionToBindSignature + 
			(hasBoundParameters ? (', ' + boundParametersAsString) : '') +
			'): '
			+ boundFunctionTypeAsString
			+ ';<br />'; 
	}
	
	return overloads;
}

function getAllBindOverloads(genericTypeNames: string[]): string {
	
	var result = '';
	
	var index: number;
	for (index = 0; index < genericTypeNames.length; ++index) {
		result += getBindOverloads(untilIndex(index, genericTypeNames));
	}
	
	return result + 'function bind(f, ...toBind: any[]) {<br />&nbsp;&nbsp;&nbsp;&nbsp;return f.bind.apply(f, [null].concat(toBind));<br />}';
}

document.write(getAllBindOverloads(range('A','F')));
```

**Comment by bryanerayner** [comment]
Can this be approached in an incremental way? What I'm interested in the use of `bind` for, in most cases, is to avoid wrapping a callback in an arrow function.

````ts

interface AngularScope
{
   $watch<T>(event:string, callback: (newValue:T, oldValue:T)=>void);
}

class Example
{
    constructor($scope: AngularScope){
        $scope.$watch('change', this.theChangeCallback.bind(this));
    }

    theChangeCallback(new: string, old:string) {
        // Do work
    }
}
````

In this circumstance, I am not using `bind` for partial application. I am using it merely to avoid writing this:

````ts
    $scope.$watch('change', (new:string, old:string)=>this.theChangeCallback(new, old));
````

What would need to be added to the compiler, just to achieve this simple integration? I feel like if `bind` were able to take on the exact type of the parent function when called with a single argument, it would be a start.

**Comment by basarat** [comment]
> `this.theChangeCallback`

@bryanerayner you should use an arrow function to define the member in the class from the get go. This is covered here : https://basarat.gitbooks.io/typescript/content/docs/tips/bind.html

**Comment by sandersn** [comment]
@bryanerayner my PR #6739 gives types to apply, call and bind. Take a look in core.d.ts: 5fe84781592a08b5294e01a2fbf42d1def07111d
I should improve the annotation to explicitly mention that `this: void` for the bound function.

**Comment by mhegazy** [comment]
this should be better now with https://github.com/Microsoft/TypeScript/pull/6739.

The ideal fix would be possible when/if https://github.com/Microsoft/TypeScript/issues/1773 is implemented.

**Comment by basarat** [comment]
@mhegazy isn't #6739 disabled (maybe only partially)? Its a bit confusing reading that + https://github.com/Microsoft/TypeScript/issues/7689 . 

A clear message about what was added will help me read it better :rose:

**Comment by mhegazy** [comment]
#6739 is merged in today. it should be available in `typescript@next` later tonight. #6739 adds two new features:
- ability to define the type of `this` argument to a function or a method. it has to be the first argument in the list.
the type specified will be checked when assigning and when calling the function.
- a new flag `--noImplicitThis`, that makes it illegal to use `this` unless defined, either explicitly in a function, e.g `function foo(this: Bar) { this } ` or implicitly, by virtue of being in a class. 

Please see https://github.com/Microsoft/TypeScript/issues/6018 for more details.

**Comment by mhegazy** [comment]
The comments in https://github.com/Microsoft/TypeScript/issues/7689 are regarding a different proposal that #6739 implemented earlier to prevent tearing off methods of classes and interfaces, as there is an implicit `this` assumption about them. that was `--strictThis` or `--strictThisChecks`

**Comment by basarat** [comment]
@mhegazy thanks! :rose:

**Comment by sebilasse** [comment]
@mhegazy and @basarat thanks for `bind`, `call` and `apply`.
How about [`name: string`] (https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Function/name) (ES6)?

**Comment by mhegazy** [comment]
this is defined in https://github.com/Microsoft/TypeScript/blob/master/lib/lib.es6.d.ts#L4201, you get that automatically when you set `--target ES6`, or if you are using `typescript@next`, you can use `--lib es6`

**Comment by sebilasse** [comment]
@mhegazy I see, but I am using it in a polyfill for `--target ES5` ;) Basically no worries, I could "extend" the `interface Function` ...

**Comment by mhegazy** [comment]
if you are using a polyfill library, it should provide a .d.ts matching what it implements, see https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/core-js/core-js.d.ts for an example.

**Comment by sandersn** [comment]
#8354 reverts the change made by #6739 because it doesn't work correctly for overloaded functions:

```ts
declare function f(): number;
declare function f(s: string): string;

// BUT TypeScript can't use the number of arguments to determine which overload `call` should use.
var n: number = f.call(null); 
var s: string = f.call(null, 'foo');
```

The compiler currently compares the last overload in declaration order, so the return type inferred for `f` is *always* `string`, never `number`.

**Comment by JoshuaKGoldberg** [comment]
> mhegazy removed this from the TypeScript 2.0 milestone on Apr 28

Oh :(

Is this issue still on the radar?

**Comment by mhegazy** [comment]
We thought we had a fix in  #6739  that solved it. but then that turned out to be invalid. please see @sandersn reply in https://github.com/Microsoft/TypeScript/issues/212#issuecomment-215488970 .

**Comment by masaeedu** [comment]
@mhegazy Does this mean the current state of overload resolution makes it too difficult to implement the feature, or can the work be salvaged? I'm trying to figure out if typed `apply`/`bind`/`call` is still on the table.

**Comment by sandersn** [comment]
#5453 needs to solve the overload problem for variadic kinds. At that point, variadic kinds plus `this` parameters should allow fully typed `apply/bind/call`. However, the prototype definitely doesn't address overload resolution and I don't think the proposal does either.

**Comment by frogcjn** [comment]
bind should return self type?

**Comment by Kovensky** [comment]
@frogcjn only if `.bind` is called with a single argument. If it's called with more than one, it changes the function signature.

**Comment by anvish** [comment]
Just FYI this is how it's handled by Flow:
```
function func(a: number, b: string) {
    console.log([a, b])
}

func.call(null, ['b'])

// 13: func.call(null, ['b'])
//     ^^^^^^^^^^^^^^^^^^^^^^ call of method `call`
// 13: func.call(null, ['b'])
//     ^^^^^^^^^^^^^^^^^^^^^^ undefined (too few arguments, expected default/rest parameters). This type is incompatible with
// 10: function func(a: number, b: string) {
//                                 ^^^^^^ string

func.call(null, ['b', 1])

// 13: func.call(null, ['b', 1])
//     ^^^^^^^^^^^^^^^^^^^^^^^^^ call of method `call`
// 13: func.call(null, ['b', 1])
//                     ^^^^^^^^ array literal. This type is incompatible with
// 10: function func(a: number, b: string) {
//                      ^^^^^^ number
```

**Comment by teppeis** [comment]
FYI: This package is a workaround for this untyped `bind()` issue.
https://github.com/teppeis/bind.ts

![bind](https://cloud.githubusercontent.com/assets/26476/25569784/01266504-2e58-11e7-8cb4-9d3b4fac63ea.png)

**Comment by tycho01** [comment]
There was a simple `bind` overload by suggested by @jcalz at https://github.com/Microsoft/TypeScript/issues/16676#issuecomment-310128938 that would already improve the current situation a bit for when only the `thisArg` is provided:
```
interface Function {
  bind<F extends Function>(this: F, thisArg: any): F
}
```

For a more general solution, I tried a `bind` attempt at https://github.com/Microsoft/TypeScript/issues/5453#issuecomment-309035811, based on that and optionally #6606:
```ts
interface Function {
    bind<
        F extends (this: T, ...args: ArgsAsked) => R,
        ArgsAsked extends any[],
        R extends any,
        T,
        Args extends any[], // tie to ArgsAsked
        Left extends any[] = DifferenceTuples<ArgsAsked, Args>,
        EnsureArgsMatchAsked extends 0 = ((v: Args) => 0)(TupleFrom<ArgsAsked, TupleLength<Args>>)
        // ^ workaround to ensure we can tie `Args` to both the actual input params as well as to the desired params. it'd throw if the condition is not met.
    >(
        this: F,
        thisObject: T,
        ...args: Args
    ): (this: any, ...rest: Left) => R;
    // ^ `R` alt. to calc return type based on input (needs #6606): `F(this: T, ...[...Args, ...Left])`
}
```

@RyanCavanaugh:
> Just a brief exploration of some of the issues here. We would need to fill in all the ?s along with descriptions of how those results would be achieved.

```
interface fn {
    (): string;
    (n: number): void;
    (n: number, m: number): Element;
    (k: string): Date;
}

var x: fn;
var y = x(); // y: string

// original version, where mutation screws everything up:
var args = [3]; // args: number[]
var c1 = x.apply(window, args); // error, not guaranteed to get a matching overload for `number[]`. had there been a `number[]` overload yielding say `Foo`, then probably infer `string | void | Element | Foo`?
args = []; // arg: still number[]
var c2 = x.apply(window, args); // c2: ditto

// without mutation, the only way around it I can see:
const args1 = [3]; // args1 type: [3], following https://github.com/Microsoft/TypeScript/issues/16389
var c1 = x.apply(window, args1); // c1: void
const args2 = []; // args2 type: for this to work this would have to be give the empty tuple type []. is it even possible to expose that?
var c2 = x.apply(window, args2); // c2: string

var b = x.bind(undefined, null); // b type: normally `(...args: Args) => fn(null, ...Args)` based on 5453 + 6606, but with `strictNullChecks` error since having `null` as the first argument would already result in no matching overloads in `fn`.
var j = b(); // j: with `strictNullChecks`: `any` since that's what `b` got due to its error. without `strictNullChecks`: `void`, as it'd just settle for the first unary overload.
```

**Comment by mmc41** [comment]
@mhegazy Any updates this. Need this for typescript to support currying, partial applications etc. in functional programming.

**Comment by Yuudaari** [comment]
People might have done this previously but I did it myself.

Generated types (supports up to 10 arguments): https://gist.github.com/Yuudaari/f4c21f8e5e3dad36e4c7f61cbddb4f22  
Generator (for any number of arguments... it's up to you!): https://gist.github.com/Yuudaari/508709f582c4aa76eeea70365d145c93

Replace native methods with ones that support these methods on class constructors: https://gist.github.com/Yuudaari/68662657485524c8f1c65250e7c2d31f  
This uses `prototype` but you can pretty easily make it not if you so desire.

Was a fun side project. I really wish there was syntactic sugar for `bind` already tho...

**Comment by tycho01** [comment]
looks like a good candidate for `lib.d.ts` :)

**Comment by Yuudaari** [comment]
@bcherny There are caveats to it; using the methods on varargs functions falls back to the provided function overloads, currently, whereas if there were just my overloads, it would not work. This also means that it can't be used for type checking, really, because the original is vague so it'll accept anything that doesn't work w/ my overloads. My overloads are only really nice for a little bit of type hinting. I'm not sure they're production ready. Also, there might be a non-negligible performance impact of adding a lot of overloads, but I haven't compared with and without.

**Comment by jcalz** [comment]
In case folks aren't aware, the tuple rest/spread work in #24897 slated for [TS 3.0](https://github.com/Microsoft/TypeScript/wiki/Roadmap#30-july-2018) will probably end up fixing this issue in the not-to-distant future.

@ahejlsberg [said](https://github.com/Microsoft/TypeScript/pull/24897#issue-194165511):
> <h3>Strong typing of bind, call, and apply</h3>
> With this PR it becomes possible to strongly type the <code>bind</code>, <code>call</code>, and <code>apply</code> methods on function objects. This is however a breaking change for some existing code so we need to investigate the repercussions.

**Comment by ffMathy** [comment]
@tycho01 very interesting. Do we know of any plans or PRs that enable us to have a fully strong-typed `bind` with an arbitrary amount of parameters?

Also, @ahejlsberg said that "we need to investigate the repercussions". Do we have a timeline on when this will happen, or an issue that keeps track of this progress?

**Comment by tycho01** [comment]
@ffMathy:

> repercussions: timeline/issue?

dunno, ask him in there!

> Do we know of any plans or PRs that enable us to have a fully strong-typed bind with an arbitrary amount of parameters?

ignoring functions with generics, a non-hacky way to do the tuple `Concat` needed for `bind` (`[...T]`) is the topic of #5453.

**Comment by ahejlsberg** [comment]
Now implemented in #27028.

**Comment by eco747** [comment]
for the original question, I think there is another solution:

just derive your class from this one

```ts
class Bindable<T> {
	bind<K extends keyof T>( name: K ) : ( ...args: EventIn<T,K> ) => void {
		return ( ...args ) => { 
			this[name as string]( ...args ); 
		}
	}
}

// just a demo class implementation
class MyClass extends Bindable<MyClass> {

    v: number;

    constructor(  value: number ) {
        super( );
        this.v = value;
   }

   // the demo method we will bind
   test( x: number ) { 
        return this.v + x;  
   }
}

let obj = new MyClass( 5 );
let fn = obj.bind( 'test' ); // here editor shows 'test' ( and 'v' and 'bind' we can remove using more elaborated code)

fn( 6 );  // here editor is waiting for a number because arguments types were propagated
fn( 'hello' ); // will fail due to bad argument type

```
